### PR TITLE
ISPN-10602 Lazily register clustered listener for counters

### PR DESCRIPTION
**Pull request by pruivo**
Only register the clustered listener when it is required. The clustered
listener has an impact on add operation performance.

https://issues.redhat.com/browse/ISPN-10602

### CODE REVIEW

File: counter/src/main/java/org/infinispan/counter/impl/manager/EmbeddedCounterManager.java [hunk 1]
```diff
@@ -320,18 +319,15 @@ private Object createCounter(String counterName) {
...
-         throw new CounterException(e);
-      }
+      notificationManager.setCache(cache());
 
       switch (configuration.type()) {
          case WEAK:
+            // topology listener is used to compute the keys where this node is the primary owner
+            // adds are made on these keys to avoid contention and improve performance
+            notificationManager.registerTopologyListener();
+            // the week counter keeps a local value and, on each event, the local value is updated (reads are always local)
```
Comment (danberindei) [inline counter/src/main/java/org/infinispan/counter/impl/manager/EmbeddedCounterManager.java:329]: Typo: week

**Review by danberindei — COMMENTED** [review]
Pedro, a bit off-topic, but what's the advantage of weak counters if writes have to update the value synchronously (via the cluster listener) on all the cluster members?

**Comment by danberindei** [comment]
I forgot one more question: have you considered also removing the counter listener when the last user listener is removed?

**Comment by pruivo** [comment]
In the weak counter, the updates happen concurrently. Each node will keep track of the keys belonging to the weak counter and only perform writes in the keys where the node is the primary owner. You save 1 round trip and you avoid lock contention as well. In addition, the reads are local (reading a field in an object)

> have you considered also removing the counter listener when the last user listener is removed?

I don't think users will remove listeners that often. I mean, adding a listener just to listen for a couple "time" and then remove it, it does not seem a frequent use case.
If you think otherwise, I can take look at it in this PR too.

**Comment by pruivo** [comment]
PR updated

**Comment by wburns** [comment]
Integrated into main, thanks @pruivo !

**Pull request merged**